perf fixes for high memory load situation

PeterSolMS opened the pull request:
- only check fragmentation when gen 2 grew and we have consumed a significant fraction of the budget
- the meaning of "significant" is modified if memory load has changed since the last gen 2 GC.
- gen 2 and UOH budgets are reduced at high memory loads.

PeterSolMS committed:
Initial version with the following features:
- only check fragmentation when gen 2 grew and we have consumed a significant fraction of the budget
- the meaning of "significant" is modified if memory load has changed since the last gen 2 GC.
- gen 2 and UOH budgets are reduced at high memory loads.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -19469,6 +19472,30 @@ bool gc_heap::init_table_for_region (int gen_number, heap_segment* region)
+    uint32_t curr_remaining_memory = 100 - memory_load;
+
+    // if the previous memory load was already at the limit, the situation cannot have gotten worse
+    if (prev_remaining_memory <= 0)
```
>   if (prev_remaining_memory <= 0) [](http://example.com/codeflow?start=1&length=34)

how can this ever be < 0?

PeterSolMS replied:
Abundance of caution - some operating system _might_ return a value > 100 for memory load - we want to treat this as 100...

Maoni0 replied:
okie but this is a uint32_t, not an int32_t :)

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -19469,6 +19472,30 @@ bool gc_heap::init_table_for_region (int gen_number, heap_segment* region)
+    float budget_fraction_consumed = budget_fraction_remaining > 0.0f ? (1.0f - budget_fraction_remaining) : 1.0f;
+    float memory_fraction_remaining = (float)curr_remaining_memory/(float)prev_remaining_memory;
+
+    return (budget_fraction_consumed >= memory_fraction_remaining);
```
I don't understand how this logic works. in most cases we can think that `memory_fraction_remaining` is always 1 because the previous and concurrent memory load are the same (they are integers so on big machines it's very likely it remains the same for a long period of time). and this method would only ever return true if `budget_fraction_consumed` is >= 1.0 which means the budget is fully consumd...

PeterSolMS replied:
Exactly right. The purpose of this method is to adapt to a _changing_ memory load. Adjustment to a _constant_ memory load is done in the budget computation.

PeterSolMS committed:
Further refinements to the logic:
- check whether gen2 has enough big blocks to fit gen1
- fix free list accounting
- in provisional mode, use majority vote to trigger compacting gen 2 GC.
- when estimating usable free list space in gen2, only look at the big blocks

danmoseley commented:
@PeterSolMS is this still active? can we close it?

stephentoub commented:
@PeterSolMS, @Maoni0, is this still being worked on?

Maoni0 commented:
we are not actively working on this currently. I'll chat with Peter what we want to do about this. likely not for 7.0.

Maoni0 commented:
it'll require quite a bit work to validate this so it'll unlikely make into 7.0. do you want us to close this for now or can it stay open till we get to it?

stephentoub commented:
Thanks, @Maoni0.  If we think it'll be reasonably soon, feel free to leave it open.  If it's just going to sit here for a long time untouched, though, let's close it for now.

Maoni0 commented:
in that case I will close this for now.
